### ISSUE TITLE
Decisions about the lib.

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/error-polyfill#26]
Several issues depend on bigger decisions about the env support, the api, what the lib should contain, etc. Before doing anything I need to examine these things and decide about them.

**Comment by inf3rno** [comment]
- I should check whether all of the environments can support the V8 API. If not (e.g. PhantomJS?), then we should decide whether we want to keep the API and give up the support by the actual environment, or we should change the API, so it won't be the V8 API any longer.
- We should decide whether the current lib should contain only the features of the V8 API, or we can add new features as well. If this will contain only  the V8 features, then we should move the extensions to a different lib, probably to a V8 extended lib or to the e3 lib.
- We should check the V8 documentation and extend it with the missing parts about class names, stack format, etc. If we keep the V8 API, then we'll need a complete documentation of it to be able to implement it on different envs.
- We should decide whether the stack parser should be on its own lib, maybe as a general self-teaching text parser or something like that.
- We should design the classes/functions of this lib before doing any modifications on it. E.g. how to prevent failing error reporters to cause infinite loops, etc. If we decide on V8 only, we'll still need a backdoor to add a custom error reporter for the extended version...

**Comment by inf3rno** [comment]
I am lucky I documented the browser differences. I did that for other devs, not for myself, but it is incredibly useful now.

**Comment by inf3rno** [comment]
**I should check whether all of the environments can support the V8 API. If not (e.g. PhantomJS?), then we should decide whether we want to keep the API and give up the support by the actual environment, or we should change the API, so it won't be the V8 API any longer.**

Let's start with this. The `error.stack` of native errors cannot be overridden by PhantomJS, because it is non-configurable. I am not sure whether this changed, I think I reported this 2 years ago. By the other browser it is possible to change it, but by native errors the `Error.prepareStackTrace` is not implemented and there is no way to inject code to the constructor. We could replace the original native classes with ouw own, but that would break compatibility with old code. I added `Error.getStackTrace(instance)` because of this. The only problem that it is not V8 compatible. So by the basics the V8 compatibility is broken. :S I need to check the examples, but I think based on this it is better to make our own API instead of using the V8 API, because we are not able to implement the V8 features in every project.

Tbh, by project which will use this lib I think the only important thing that have an interface, which works in every environment the exact same. It does not matter whether it is the V8 Error API or a custom one. So it is just by jquery. I think the error-polyfill is not the proper project name, it should be jerror or something like that.

**Comment by inf3rno** [comment]
**We should decide whether the current lib should contain only the features of the V8 API, or we can add new features as well. If this will contain only the V8 features, then we should move the extensions to a different lib, probably to a V8 extended lib or to the e3 lib.**

The previous point answer this question, it is better to have just a single lib.

**Comment by inf3rno** [comment]
**We should check the V8 documentation and extend it with the missing parts about class names, stack format, etc. If we keep the V8 API, then we'll need a complete documentation of it to be able to implement it on different envs.**

We won't keep the V8 API, but we could use some parts of it, e.g. the structure of the stack object or the function names.

**Comment by inf3rno** [comment]
**We should decide whether the stack parser should be on its own lib, maybe as a general self-teaching text parser or something like that.**

I am not sure whether the problem can be generalized. If so, we can create a separate lib. I think we should first implement it, and decide only after that whether it is worth the effort to make a separate repo for it.

**Comment by inf3rno** [comment]
**We should design the classes/functions of this lib before doing any modifications on it. E.g. how to prevent failing error reporters to cause infinite loops, etc. If we decide on V8 only, we'll still need a backdoor to add a custom error reporter for the extended version...**

I think this will be the hard part, so I split this up into multiple comments.

**Comment by inf3rno** [comment]
Let's check what features we have. Just only the V8 features for now.

`e.stack | e.stacktrace` -> returns the stack string of the error object
`Error.captureStackTrace(e, terminator)` -> sets the present stacktrace on the error object above the terminator (the terminator calls the function directly or indirectly)

e.g.

```js
var UserError = function (message){
    this.name = "UserError";
    this.message = message;
    Error.captureStackTrace(this, this.constructor);
};
```

`Error.prepareStackTrace = f(e, frames , warnings)` -> transforms the stack object into a custom stack string

```js
Error.prepareStackTrace = function (throwable, frames, warnings) {
    var string = "";
    string += throwable.name || "Error";
    string += ": " + (throwable.message || "");
    if (warnings instanceof Array)
        for (var warningIndex in warnings) {
            var warning = warnings[warningIndex];
            string += "\n   # " + warning;
        }
    for (var frameIndex in frames) {
        var frame = frames[frameIndex];
        string += "\n   at " + frame.toString();
    }
    return string;
};
```

`Error.stackTraceLimit = int` -> sets the stack trace limit

**Comment by inf3rno** [comment]
```js
var lib = require("lib"),
    UserError = lib.UserError;
// native (better to wrap):
    try {
        smellingCode();
    } catch (e){
        e = lib(e);
        // wraps the native error, e.g. SyntaxError -> SyntaxErrorWrapper extends SyntaxError 
        if (e instanceof SyntaxError)
            ...
        else if (e instanceof UserError)
            ...
    }
    e.name
    e.message
    e.stack -> returns the stack object
    e.stack.throwable = e
    e.stack.frames[]
    e.stack.warnings[]
    e.stack.toString() -> returns the stack string
// custom:
    MyError extends lib.UserError -> not sure whether it is possible with custom constructor
// lib functions:
    lib.on("uncaughtError", function (e){})
    lib.Stack -> overriding the Stack.prototype.toString will do the same as prepareStackTrace
    lib.captureStack(terminator) -> returns the present Stack object prior the terminator call
```

**Comment by inf3rno** [comment]
advanced features:

```js
lib.on("error", function (e){}) // -> fires when a new error object is created or a native error object is wrapped, good for logging errors
```

**Comment by inf3rno** [comment]
By on.error and on.uncaughtError we should keep in mind that multiple versions of this lib can run in the same process and other libs can listen these event handlers as well. So we need to somehow connect these multiple versions and use some kind of no-conflict mode with the other libs. Both will be hard to implement. Probably we need to set some properties to non-configurable (e.g. Error.prepareStackTrace) or add a setter for them to be compatible with multiple libs. I am not sure whether it is possible to solve this, we'll see. The multi-version part is solvable I think, we just need a global point, where we can register different versions of the lib, e.g. Error.lib.registry or something like that. Backward compatibility is guaranteed only by the same major version, so this is necessary if we have later a 2.0.

**Comment by inf3rno** [comment]
I think we should check common libs:

 - jquery uses the event handler, it does not override anything: https://api.jquery.com/error/ 
 - angular: only the angular docs overrides window.onerror with some custom stuff which transforms the stack and sends it to googly analytics. the lib itself calls window.onerror(args), so it is not a problem
 - react: jest uses window.addEventListener("error", ...), I think that is a testing lib. by guardedCallback react uses the same approach temporarily, so removeEventListener too. otherwise it does not care much about error handling.

I think most of the libs let the user do the error handling stuff and don't override the observers with their own implementation, which is good for us.

**Comment by inf3rno** [comment]
Global error handling has different versions. It started with `window.onerror(message, source, lineno)`, continued with `window.onerror(`message, source, lineno, colno)` and ended with `window.onerror(`message, source, lineno, colno, error)`. Meanwhile they added `window.addEventListener("error", f(event){})` and the ErrorEvent contains the same properties as the arguments of the onerror.  Meanwhile msie added the `window.attachEvent("onerror", f(){})`, which is not that well documented. I found a gist, which uses a workaround https://gist.github.com/dimsemenov/1d79ef4af0fdf73d7de6 by getting the error event from `window.event` in the handler and does not try to construct its own event object. I don't think I'll follow that approach, but who know. Afaik. `window.event` is deprecated, or at least nobody uses it.

**Comment by inf3rno** [comment]
Edge don't have error object support? Wtf? https://blog.sentry.io/2016/01/04/client-javascript-reporting-window-onerror

**Comment by inf3rno** [comment]
It is possible to wrap event listeners in try-catch blocks, which is good if we want to have stack by old browsers too, but using try-catch in those will seriously slow down the code in an already slow environment.

**Comment by inf3rno** [comment]
There is a document here about how to extend Firefox DOM nodes with custom code, so it is possible to override addEventListener in those. https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript-DOM_Prototypes_in_Mozilla You can do the same by setTimeout, etc. So in theory it is possible to add try-catch blocks to every event handler and catch the error object instead of just the message. In practice it does not matter.

**Comment by inf3rno** [comment]
Another approach would be to keep the error objects logged in the on.error for a few secs. Hmm I don't think the `lib.on("error"` is anything useful. Nobody wants to log every error objects, just the uncaught ones. But I can use the constructor to keep the errors for 1-2 secs in an array, so even in msie8 I would be able to get the stack of the uncaught error using this array. It would work, and it would be faster than using huge try-catch blocks. I think instead of `on("error"` it would be more beneficial to use some kind of bus. Otherwise somebody will assume that throwing the error is preventable somehow. I need to think more about this.

**Comment by inf3rno** [comment]
Before starting a new lib I'll add the stack parser to this one. The new lib will have almost the same features, so I'll just need to rename some of the functions.

**Comment by inf3rno** [comment]
I was thinking on the API again. The problem with the current lib is that `err.stack` does not work, because I cannot do constructor injection by native errors. By v8 there is `Error.prepareStackTrace`, which helps only by the string conversion, but the `err.stack` will still result in a string. By PhantomJS as far as I remember you cannot override `err.stack`. By opera you have `err.stacktrace`. So that's all we currently have. I decided because of this to define a new method to get the `Stack stack` instead of just serving `String stack`. Another cause was lazy load. I don't parse the stack until somebody calls `Error.getStackTrace`. I think this is fine, because we can work with native errors the same way as we usually do, and if we need a stack object, then we call the new method. Even by v8 there is no other way, so this is an extra feature. By Opera I need to check whether it is possible to remove the `stacktrace` property and add a `stack` property with `Object.defineProperty`. If it is possible by MSIE, Firefox and Opera, then I can try to do something about it, but I think it is better to leave that property alone. I think I decided this way because you'd need to call some kind of function before accessing the stack string. For example: `e(err); console.log(err.stack);`. This is a lot easier to forget than `console.log(Error.getStackTrace(err))`, because it is 2 steps instead of just one.

**Comment by inf3rno** [comment]
I am not sure about this. After I am done with the new stack implementation and the other stuff, I'll think about moving the project to a new lib, with a custom (non-v8) API again. Getting the stack in the current one is not v8, but in v8 I was able to get only the stack string, so it is more or less compatible. It is still better to have a native error object instead of a wrapped one, since `e.constructor == Error` and `e1 === e2` works, but I can support the `wrap(e) instanceof Error` part, which is the most important one, and the wrapped error could have `w.stack`, which is much better than using `Error.getStackTrace(e)`. The drawback is IE 10+ compatibility, because I'll need to use getters to lazy load the stack. Hmm after all I don't really need getters if I access stack frames with methods instead of properties, something like `e.stack.getFrames()`. I need to think about this and wait whether it is possible to support IE 6-9.

**Comment by inf3rno** [comment]
The current lib doesn't even support having a stack object, it just returns a stack string. Ofc. v8 does the same. It would be better to get an object with a toString method, but I guess prepareStackTrace does not allow that.

**Comment by inf3rno** [comment]
It is possible to capture the frames and export it outside by v8 and keep the custom prepareStackTrace:
```js
var defPS = function (throwable, frames, warnings) {
    var string = "";
    string += throwable.name || "Error";
    string += ": " + (throwable.message || "");
    if (warnings instanceof Array)
        for (var warningIndex in warnings) {
            var warning = warnings[warningIndex];
            string += "\n   # " + warning;
        }
    for (var frameIndex in frames) {
        var frame = frames[frameIndex];
        string += "\n   at " + frame.toString();
    }
    return string;
};


var x = Error.prepareStackTrace;
var m = new WeakMap();
Object.defineProperties(Error, {
    prepareStackTrace: {
        set: function (v){
            x = v;
        },
        get: function (){
            var q = (x instanceof Function)?x:defPS;
            return function (throwable, frames){
                m.set(throwable, frames);
                return q.apply(this, arguments);
            };
        }
    }
});

var e = new Error("asdfg");
e.stack; // lazy load
console.log(m.get(e));
```

The only drawback of this solution, that the prepareStackTrace will be set by default, so codes which check that will break, but codes, which just set the property will work fine. I think we can pay this price. Still better than parsing v8 from string, especially if somebody sets a custom string... Hmm still I don't think I'll support json generation by this lib. It will return only the string and nothing more.

**Comment by inf3rno** [comment]
Ok. So the current lib works this way:

v8:
```
Error.getStackTrace(err) || Error.captureStackTrace(err)
err.stack // lazy load
Error.prepareStackTrace(err, frames) -> stackString
stackString
```

non-v8
```
calibrateStackStringParser()
Error.getStackTrace(err)
source.getFrameStrings(err) | source.captureFrameStrings(paddingSize) -> frameStrings[], functionReferences[]
parser.getFrames(frameStrings,functionReferences) -> frames[]
Error.prepareStackTrace(err, frames) -> stackString
stackString
```

**Comment by inf3rno** [comment]
I need example stacks from different environments. It would be fun to parse stacks from other languages as well. :-) If that is possible, then I think the js code will not cause any trouble.

**Comment by inf3rno** [comment]
The V8 StackTrace API does not contain function arguments, but in some environments those are part of the function calls. Firefox appear to add primitives. The old Opera appear to copy the actual line to the second frame line. The first frame line contains, whether the error is from the html or from a linked js file. I have to prepare the lib for that. I am not sure how. I might need to install the old Opera to check it. Data URI probably won't work, but maybe iframes do.

Another option to find the words in the frame template and replace each one with `\S+` if they don't match an actual error. For example:

```
"Line 15 of inline#1 script in file://localhost/G:/js/test/functional/testcase1.html\n"
-> Line @{line} of inline#1 script in @{path}\n"
"Line 18 of linked script file://localhost/G:/js/stacktrace.js\n" +
-> does not match
-> .* @{line} of inline#1 script in @{path}\n"
-> still no match
-> .* @{line} .* @{path}\n"
-> match
-> Line @{line} .* @{path}\n"
-> match
-> Line @{line} of .* @{path}\n"
-> match
-> Line @{line} of inline#1 .* @{path}\n"
-> does not match
-> Line @{line} of .* script .* @{path}\n"
-> match
-> Line @{line} of .* script in @{path}\n"
-> does not match
"Line 15 of inline#1 script in file://localhost/G:/js/test/functional/testcase1.html\n"
-> Line @{line} of .* script .* @{path}\n"
-> match -> we can modify the template
```
We need a negative test too, which should not match otherwise using `.*` will work just fine...  Hmm probably we don't need a negative test, we just need to check whether the `line` and `path` are properly parsed. If they are not, then the pattern is too broad. I think this part won't be smart enough to always work. Maybe we can change the `.*` to `\S+` by these words if it matches too, so it won't be too broad. So we need to look for variable regions of the stack string. We won't know what these are, but we can save them as well I think. Hmm maybe I could use the same algorithm to find variable regions as I use for the probes. The only difference here that I don't know what the actual value means by these regions. They are problematic btw if they are close to other variable regions for example here `Line @{line} of .* script .* @{path}` the second `.*` is close to the `path`, which is a variable region and can contain even whitespace in other environments. So there is no way to decide where the unknown variable string ends and where the known variable path starts. I need to find a solution for this problem. The parser must be smart enough to find the beginning of the path based on previous examples. So it should compare previous paths, which is a little bit complicated, because there can be paths like `eval at window.onload (x.html:21), <anonymous>:3:7` or `Unknown script code:8:2`. The function name can contain whitespace as well e.g. `Global code`, or by Opera it is the actual line. So in theory there is no way to properly separate these variable regions from the actual path without understanding what the words mean.

**Comment by inf3rno** [comment]
Another topic we haven't discussed is using multiple frame strings. Sometimes there is no function name by chrome and v8: `at @{path}:@{line}:@{col}`. We need to probe these situations as well, otherwise the `at @{func} (@{path}:@{line}:@{col})` won't match. By the probing period it should be possible to add new frame templates and find out whether they are mutually exclusive or whether they both match, so we should use a priority order. After the probing period we can detect the variable regions, but since we cannot probe them, there won't be a way to find out whether we are talking about a new frame template or we should use the same template and add a variable region. I think it will be hard to adapt the existing templates sometimes, but I don't think there is a way to build a new template from scratch without probing. That would require a true AI or even an AI won't be able to solve this without knowing the language and the meaning of a stack string.

**Comment by inf3rno** [comment]
The upper means that the V8 API is not general enough. I am still thinking on using a different API and this is a pro towards that direction.

**Comment by inf3rno** [comment]
Another think in favor of using a different API that there is no way to reach the stack object by V8. The whole `Error.getStackTrace` thing is about this. I mean the best way would be to return a stack object with a toString method by getting the error.stack directly, but not even the V8 gives an obvious solution to do that. Not to mention the other libs. The phantomjs throws an error if you try to touch native error properties... I added the `Error.getStackTrace(e)` because the V8 does not have a feature, which returns a stack object, so we can call this a new feature, while the stack string generation a well documented already existing feature. Sooner or later we will find other new features and the Error class won't be enough. I don't like polluting global NS, so instead of writing a polyfill, we need a shim with a separate API. The only problem that it is a lot more convenient to write `e.stack` instead of `mylib(e).stack`. We can easily forget the mylib part and end up with the original stack string. If we check it in node only, we might don't even recognize the error.

**Comment by inf3rno** [comment]
Which makes me think this twice that `new UserError().stack` looks so cool by e3. It is not a native error, and I use it most of the times, that's why it is hard to give it up.

**Comment by inf3rno** [comment]
- I'll override the `Error.prototype.stack` and the `Error.prototype.stacktrace` variables and remove this `getStackTrace` method. If the phantomjs fails, then I won't support the phantomjs. All that matters that the real browsers don't fail...

**Comment by inf3rno** [comment]
I might move the whole code to a separate lib and let the polyfill depend on that lib. That way I have both solutions. The polyfill will cover only the `captureStackTrace`, `prepareStackTrace`, `stackTraceLimit`, `Error.prototype.stack` parts. So the uncaught exceptions, etc. should be part of the lib instead of this repo. I think the lib should not be the e3. Another option to add polyfill to e3 and do something like `require("e3").polyfill()`. But I'd rather avoid that. Hmm actually I could use an e3 submodule, which contains only these things, but it is cleaner to move the whole thing to a different repo.

**Comment by inf3rno** [comment]
I'll need this dependency in both e3 and this polyfill. I don't want the e3 to depend on this polyfill. It should work in no-conflict mode too. Not to mention that  the dependency would support phantomjs too, but this polyfill won't support it.

**Comment by inf3rno** [comment]
I don't want to depend on the u3 and o3 libs. I don't want to use es6 features either. I am not sure about the capability lib.

**Comment by inf3rno** [comment]
I think this dependency will be in a namespaced npm repo. The errors, and such names are taken.

**Comment by inf3rno** [comment]
As far as I understand shims are an abstraction layer over different APIs, e.g.
```js
function on(type, listener) {
	if (window.addEventListener)
		this.addEventListener(type, listener, false);            
	else if (window.attachEvent)
		this.attachEvent("on"+type, listener);            
	else
		this["on"+type] = listener;           
}
```

Polyfills can be similar, but they use to implement a well defined usually standardized modern API.
```js
if (!window.addEventListener)
	function addEventListener(type, listener, blah) {
		if (window.attachEvent)
			this.attachEvent("on"+type, listener);            
		else
			this["on"+type] = listener;           
	}
```

I think that's the main difference, so the repo can have the "error-shim" name or something like that. It does not really matter though.

**Comment by inf3rno** [comment]
I think the "error-shim" name is ok. It is not taken in NPM. I need to think about namespaced NPM repos again. I might use namespaces for utility repos and publish high level repos without them. So for example the polyfill and e3 can be published without namespace, while the shim should be namespaced.

**Comment by inf3rno** [comment]
I created a repo https://github.com/inf3rno/error-shim, I'll copy the whole code to it, and I'll add it as dependency after it reached 0.1.x.
